batchflow/models/tf/xception.py

```diff
@@ -0,0 +1,199 @@
+                    x = cls.block(x, name='block-'+str(i), **args)
+
+            # Middle flow: thorough processing
+            with tf.variable_scope('middle_flow'):
+                middle_stages = middle.pop('num_stages', 0)
+                for i in range(middle_stages):
+                    args = {**kwargs, **middle, **cls.slice_dict(middle, i)}
+                    x = cls.block(x, name='block-'+str(i), **args)
+
+            # Exit flow: final increase in number of feature maps
+            with tf.variable_scope('exit_flow'):
+                exit_stages = exit.pop('num_stages', 0)
+                for i in range(exit_stages):
+                    args = {**kwargs, **exit, **cls.slice_dict(exit, i)}
+                    x = cls.block(x, name='block-'+str(i), **args)
+        return x
+
+
+    @classmethod
+    def block(cls, inputs, filters, combine_type='sum', name='block', **kwargs):
+        """ Basic building block of the architecture.
+        For details see figure 5 in the article.
+
+        Parameters
+        ----------
+        filters : sequence of 3 ints
+            Number of feature maps in each layer
+        """
+        strides = (1, kwargs.pop('strides', 1), 1)
+        x = inputs
+
+        with tf.variable_scope(name):
+            # Prepare skip-connection
+            if combine_type == 'sum':
+                shortcut = inputs
+            elif combine_type == 'conv':
+                shortcut = conv_block(inputs, 'cn', filters[-1], kernel_size=1,
+                                      strides=strides[1], name='shortcut')
+
+            # Three consecutive separable blocks
+            for i, filter_ in enumerate(filters):
+                x = cls.separable_block(x, filter_, strides=strides[i],
+                                        name='separable_conv-{}'.format(i), **kwargs)
+
+            outputs = tf.add_n([x, shortcut])
+        return outputs
+
+
+    @classmethod
+    def separable_block(cls, inputs, filters, kernel_size=3, strides=1, rate=1,
+                        depth_activation=False, name='separable_block', **kwargs):
+        """ Separable convolution, followed by pointwise convolution with batch-normalization in-between. """
+        layout = 'nacna' if depth_activation else 'ncn'
+        data_format = kwargs.get('data_format')
+
+        with tf.variable_scope(name):
+            x = cls._depthwise_conv(inputs, kernel_size=kernel_size, strides=strides, dilation_rate=rate,
+                                    data_format=data_format, padding='same', name='depthwise')
+
+            x = conv_block(x, layout, filters=filters, kernel_size=1, **kwargs)
+        return x
+
+
+    @classmethod
+    def _depthwise_conv(cls, inputs, depth_multiplier=1, name='depthwise_conv', **kwargs):
+        data_format = kwargs.get('data_format')
+        filters = cls.num_channels(inputs, data_format)
+        depthwise_conv = tf.keras.layers.SeparableConv2D(filters=filters,
```

Comment:
This only works with 2d data, but we definitely need Nd solution. There is separable conv in convblock. Does it fit the purpose?

Reply:
The reason I need separate function for `depthwise` conv is to insert other layers (namely, batch-norm and activation) between depthwise and pointwise convolutions. Current implementation of separable convolution does not allow to do so;
Furthermore, `keras` implementation is 2 times faster than the previous one at build time, and 35% faster at train time.
I can make `depthwise_conv` function inside conv.py, which would look pretty similar to `conv` function: get number of dimensions and choose appropriate layer to apply